pkg/inputs/mysqlbatch/input.go

```diff
@@ -365,86 +373,95 @@ func (plugin *mysqlBatchInputPlugin) waitFinish(positionCache position_store.Pos
 	}
 }
 
-func InitTablePosition(db *sql.DB, positionCache position_store.PositionCacheInterface, tableDef *schema_store.Table, scanColumn string, estimatedRowCount int64) (bool, error) {
+func InitTablePosition(
+	db *sql.DB,
+	positionCache position_store.PositionCacheInterface,
+	tableDef *schema_store.Table,
+	scanColumns []string,
+	estimatedRowCount int64) (bool, error) {
+
 	fullTableName := utils.TableIdentity(tableDef.Schema, tableDef.Name)
-	max, _, exists, err := GetMaxMin(positionCache, fullTableName)
+	_, _, exists, err := GetMaxMin(positionCache, fullTableName)
 	if err != nil {
 		return false, errors.Trace(err)
 	}
 
 	if !exists {
-		log.Infof("[InitTablePosition] init table position")
-
-		var scanType string
-		if scanColumn == "*" {
-			maxPos := TablePosition{Column: scanColumn, Type: PlainInt, Value: 1}
-			minPos := TablePosition{Column: scanColumn, Type: PlainInt, Value: 0}
-			if err := PutMaxMin(positionCache, fullTableName, maxPos, minPos); err != nil {
+		maxPositions := make([]TablePosition, len(scanColumns))
+		minPositions := make([]TablePosition, len(scanColumns))
+		if IsScanColumnsForDump(scanColumns) {
+			maxPositions[0] = TablePosition{Column: ScanColumnForDump, Type: PlainInt, Value: 1}
+			minPositions[0] = TablePosition{Column: ScanColumnForDump, Type: PlainInt, Value: 0}
+			if err := PutMaxMin(positionCache, fullTableName, maxPositions, minPositions); err != nil {
 				return false, errors.Trace(err)
 			}
+			log.Infof("[InitTablePosition] scan dump PutMaxMin table: %v, maxPos: %+v, minPos: %+v", fullTableName, maxPositions, minPositions)
 		} else {
-			max, min := FindMaxMinValueFromDB(db, tableDef.Schema, tableDef.Name, scanColumn)
-			maxPos := TablePosition{Value: max, Type: scanType, Column: scanColumn}
-			minPos := TablePosition{Value: min, Type: scanType, Column: scanColumn}
-			if err := PutMaxMin(positionCache, fullTableName, maxPos, minPos); err != nil {
+			for i, column := range scanColumns {
+				max, min := FindMaxMinValueFromDB(db, tableDef.Schema, tableDef.Name, column)
```

Comment:
should query all pk in one shot, otherwise for non-first pks, it would be a full table scan I think.

Reply:
changed